### PR TITLE
fix: improve parsing of `--env` flag

### LINKED ISSUE #2642: --env foo="" is true instead of being "".
**Describe the bug**

--env foo="" is true instead of being "".

**To Reproduce**

1. Call webpack with the option `--env foo=""`
2. In webpack.config.js foo is equal to true instead of "".

**Expected behavior**

Foo should be equal to "" instead of true.

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
Yes

**If relevant, did you update the documentation?**
will add an example in webpack.js.org after merge.

**Summary**

Fixes #2642 

**Does this PR introduce a breaking change?**
Maybe

**Other information**
No

### CODE REVIEW

File: packages/webpack-cli/lib/webpack-cli.js [hunk 1]
```diff
@@ -373,8 +373,11 @@ class WebpackCLI {
             {
                 name: 'env',
                 type: (value, previous = {}) => {
+                    // for https://github.com/webpack/webpack-cli/issues/2642
+                    const regExpForSplitting = (value.match(/=/g) || []).length === 1 ? /=/ : /=(.+)/;
```
Comment (anshumanv) [inline packages/webpack-cli/lib/webpack-cli.js:377]: why?

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: When passing `--env foo=""` via terminal the value is actually `foo=` and not `foo=""` hence setting `env.foo=` = `true`,

I don't think it is problem with our CLI, the terminal just ignores `""`.

To get the desired behavior we need to pass `--env foo=\"\"` with this the value is `foo=""` and hence the desired behavior.

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: I am not sure if we should accept this change it is more of a workaround for the situation. Maybe we can mention this somewhere in docs and leave it as is.

File: packages/webpack-cli/lib/webpack-cli.js [hunk 2]
```diff
@@ -389,7 +392,11 @@ class WebpackCLI {
                         }
 
                         if (index === splitKeys.length - 1) {
-                            prevRef[someKey] = val || true;
```
Comment (anshumanv) [inline packages/webpack-cli/lib/webpack-cli.js:392]: @snitin315 bug is here, I don't think you need diff regex, can you verify? 

what's the result with original regex and this removed?

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:392]: Screenshot 2021-04-18 at 3 28 48 PM

Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:392]: hmm yes it won't help, lemme thing of something else

**Review by rishabh3112 — APPROVED** [review]
lgtm

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: There is logical error, for any value containing = this will match, better to look for last char in the string and if it's `=` then set value as `""`

**Review by anshumanv — CHANGES_REQUESTED** [review]

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: Can you give example?

**Comment by rishabh3112** [comment]
> There is logical error, for any value containing = this will match, better to look for last char in the string and if it's = then set value as ""

Yup, good catch. One more thing we can do is to introduce some special character for empty string like we do in parser grammars. ``` ` ``` could be tried for this.

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: yep, look how it uses first regex in all cases

image

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: For any value containing only one `=` it will match and the corresponding regexp will parse it correctly. I believe test cases are already in place. Am I missing something here?

Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: maybe we are good with this, I'm not sure why the other was used, but there was some suggestion on the PR, try with this regex in all cases, we might be good with this (`/=/`)

Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: i.e remove the regexExpForSplitting logic

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: I will improve and add more tests 👍

Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: 💯 💯 💯

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:377]: Other is for the case of multiple `=`. 

`--env foo=bar=app` => `foo: 'bar=app'`

Comment (anshumanv) [reply packages/webpack-cli/lib/webpack-cli.js:377]: hmm yes, what about cases when `foo=bar=""` quite absurd but possibility

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js]: Updated 👍🏻

File: packages/webpack-cli/lib/webpack-cli.js [hunk 3]
```diff
@@ -373,6 +373,11 @@ class WebpackCLI {
             {
                 name: 'env',
                 type: (value, previous = {}) => {
+                    // for https://github.com/webpack/webpack-cli/issues/2642
+                    if (value.endsWith('=')) {
+                        value.concat('""');
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:378]: Something weird with coverage here

**Comment by alexander-akait** [comment]
Added more tests + fix edge cases with `--verison=` (empty string)

**Review by alexander-akait — APPROVED** [review]

**Comment by alexander-akait** [comment]
After CI will be green I want to merge it, because we need finish https://github.com/webpack/webpack-cli/pull/2632 and when finish testing on webpack-dev-server@4, after this we need do release (and webpack-dev-server release too)

**Review by anshumanv — APPROVED** [review]
